sbncode/Calibration/TrackCaloSkimmer_module.cc

```diff
@@ -0,0 +1,460 @@
+
+  // Track - associated data
+  art::FindManyP<recob::Track> fmTracks(PFParticleList, e, fTRKproducer);
+  art::FindManyP<recob::Hit, recob::TrackHitMeta> fmtrkHits(tracks, e, fTRKproducer);
+  art::FindManyP<anab::Calorimetry> fmCalo(tracks, e, fCALOproducer);
+
+  // Collect raw digits for saving hits
+  std::vector<art::Ptr<raw::RawDigit>> rawdigitlist;
+  for (const art::InputTag &t: fRawDigitproducers) {
+    art::ValidHandle<std::vector<raw::RawDigit>> thisdigits = e.getValidHandle<std::vector<raw::RawDigit>>(t);
+    art::fill_ptr_vector(rawdigitlist, thisdigits);
+  }
+
+  const geo::GeometryCore *geometry = lar::providerFrom<geo::Geometry>();
+
+  // The raw digit list is not sorted, so make it into a map on the WireID
+  std::map<geo::WireID, art::Ptr<raw::RawDigit>> rawdigits;
+  for (const art::Ptr<raw::RawDigit> &d: rawdigitlist) {
+    
+    std::vector<geo::WireID> wids;
+    // Handle bad channel ID
+    try {
+      wids = geometry->ChannelToWire(d->Channel());
+    }
+    catch(...) {
+      continue;
+    }
+
+    // Ignore channel with no mapped wire
+    if (wids.size() == 0) continue;
+
+    // Ignore wires that are already mapped
+    if (rawdigits.count(wids[0])) continue;
+
+    rawdigits[wids[0]] = d;
+  } 
+
+  // service data
+
+  // Build global track info
+  std::vector<GlobalTrackInfo> track_infos;
+  for (const recob::Track &t: *tracks) {
+    track_infos.push_back({
+      t.Start(), t.End(), t.StartDirection(), t.EndDirection()
+    });
+  }
+
+  for (art::Ptr<recob::PFParticle> p_pfp: PFParticleList) {
+    const recob::PFParticle &pfp = *p_pfp;
+
+    const std::vector<art::Ptr<recob::Track>> thisTrack = fmTracks.at(p_pfp.key());
+    if (thisTrack.size() != 1)
+      continue;
+
+    art::Ptr<recob::Track> trkPtr = thisTrack.at(0);
+
+    std::vector<art::Ptr<anab::Calorimetry>> emptyCaloVector;
+    const std::vector<art::Ptr<anab::Calorimetry>> &calo = fmCalo.isValid() ? fmCalo.at(trkPtr.key()) : emptyCaloVector;
+
+    std::vector<art::Ptr<recob::Hit>> emptyHitVector;
+    const std::vector<art::Ptr<recob::Hit>> &trkHits  = fmtrkHits.isValid() ? fmtrkHits.at(trkPtr.key()) : emptyHitVector;
+
+    std::vector<const recob::TrackHitMeta*> emptyTHMVector;
+    const std::vector<const recob::TrackHitMeta*> &trkHitMetas = fmtrkHits.isValid() ? fmtrkHits.data(trkPtr.key()) : emptyTHMVector;
+
+    float t0 = std::numeric_limits<float>::signaling_NaN();
+    if (fmT0.isValid() && fmT0.at(p_pfp.key()).size()) t0 = fmT0.at(p_pfp.key()).at(0)->Time();
+
+    if (fRequireT0 && fmT0.at(p_pfp.key()).size() == 0) {
+      continue;
+    }
+
+    std::cout << "Processing new track! ID: " << trkPtr->ID() << " time: " << t0 << std::endl;
+
+    FillTrack(*trkPtr, pfp, t0, trkHits, trkHitMetas, calo, rawdigits, track_infos);
+  }
+}
+
+// helpers
+float HitMinTime(const std::vector<sbn::HitInfo> &hits, bool TPCE) {
+  double min = -1;
+
+  for (const sbn::HitInfo &h: hits) {
+    bool hit_is_TPCE = h.tpc <= 1;
```

Comment:
<=1 ? So, either 0 or 1? Is that right?

Reply:
Yeah. I'd be happy to switch this out for a call to the Geometry service if such a functionality exists.

Reply:
(see below comment on adding a comment to help document)